### PR TITLE
[3658] Finalize language switcher implementation

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/3658

### CODE REVIEW

**Comment by aachavarria** [comment]
@rart Should I assume that the locales are already there? and avoid to use the async import?

File: templates/web/site-config.ftl [hunk 1]
```diff
@@ -48,6 +48,7 @@
     <script src="/studio/static-assets/libs/js-cache/cache.js?version=${UIBuildId!.now?string('Mddyyyy')}"></script>
 
     <script>
+        $('html').attr('lang', CStudioAuthoringContext.lang);
```
Comment (rart) [inline templates/web/site-config.ftl]: Write it directly on the tag.

```ftl
<html lang=“${ftlVarWhereValue}”>
```

File: ui/app/src/components/CrafterCMSNextBridge.tsx [hunk 2]
```diff
@@ -23,11 +23,19 @@ import ThemeProvider from '@material-ui/styles/ThemeProvider';
...
-export const intl = createIntl({
-  locale: 'en',
-  messages: {}
-}, cache);
+export async function _createIntl(locale: any) {
+  if(!locale) locale = 'en';
+  return createIntl(
+      {
+        locale,
+        messages: await import(`../translations/locales/${locale}.json`),
```
Comment (rart) [inline ui/app/src/components/CrafterCMSNextBridge.tsx]: let’s discuss how to make this without delaying render

File: ui/app/src/components/CrafterCMSNextBridge.tsx [hunk 3]
```diff
@@ -23,11 +23,23 @@ import ThemeProvider from '@material-ui/styles/ThemeProvider';
...
 
 const cache = createIntlCache();
+const locale = document.documentElement.lang;
+export const intl = _createIntl(locale);
 
-export const intl = createIntl({
-  locale: 'en',
-  messages: {}
-}, cache);
+export async function _createIntl(locale: any) {
```
Comment (rart) [inline ui/app/src/components/CrafterCMSNextBridge.tsx]: Type param

File: ui/app/src/components/CrafterCMSNextBridge.tsx [hunk 4]
```diff
@@ -23,11 +23,23 @@ import ThemeProvider from '@material-ui/styles/ThemeProvider';
...
+  if (!locale) {
+    locale = 'en';
+    // TODO: throw console error
+  }
+  
+  return createIntl(
+    {       
+      locale,
+      messages: await import(`../translations/locales/${locale}.json`),
```
Comment (rart) [inline ui/app/src/components/CrafterCMSNextBridge.tsx]: Let’s discuss who to not delay rendering

Comment (aachavarria) [reply ui/app/src/components/CrafterCMSNextBridge.tsx]: yep

File: templates/web/site-config.ftl (hunk 1 above)
Comment (rart) [reply templates/web/site-config.ftl]: If not an FTL var, set it in studio-context.ftl where the CstudioAuthoringContext var is populated in the first place.

File: ui/app/src/components/CrafterCMSNextBridge.tsx [hunk 5]
```diff
@@ -21,12 +21,31 @@ import React, { Suspense } from 'react';
...
+import EN from '../translations/locales/en.json'
+import ES from '../translations/locales/es.json'
+import DE from '../translations/locales/de.json'
+import KO from '../translations/locales/ko.json'
 
 const cache = createIntlCache();
+const locale = document.documentElement.lang;
+
+function selectMessages() {
+  if(locale === 'en') {
```
Comment (rart) [inline ui/app/src/components/CrafterCMSNextBridge.tsx]: Use a switch statement.

File: ui/app/src/utils/auth.ts [hunk 6]
```diff
@@ -27,8 +27,18 @@ export function setRequestForgeryToken() {
   setGlobalHeaders({'X-XSRF-TOKEN': token});
 }
 
+export function getCookie(name:string){
```
Comment (rart) [inline ui/app/src/utils/auth.ts]: These are more generic than auth. Don’t belong here.

File: ui/app/src/components/CrafterCMSNextBridge.tsx [hunk 7]
```diff
@@ -17,19 +17,72 @@
...
-export const intl = createIntl({
-  locale: 'en',
-  messages: {}
-}, cache);
+export let intl = getIntl(getCurrentLocale());
+
+// @ts-ignore
+document.addEventListener('setlocale', (e: CustomEvent<string>) => {
+  intl = getIntl(e.detail);
+  updateIntl(intl);
```
Comment (rart) [inline ui/app/src/components/CrafterCMSNextBridge.tsx:42]: Add this function to codebase-bridges.tsx

```js
export function updateIntl(nextIntl: IntlShape) {
  // @ts-ignore
  if (window.CrafterCMSNext) {
    // @ts-ignore
    window.CrafterCMSNext.i18n.intl = nextIntl;
  }
}
```

File: ui/app/src/utils/auth.ts (hunk 6 above)
Comment (aachavarria) [reply ui/app/src/utils/auth.ts]: not needed anymore

**Review by rart — APPROVED** [review]

**Pull request merged**